### ISSUE TITLE
Refactor Gravsearch processing logic in SearchResponder

### ISSUE THREAD

**Issue by tobiasschweizer** [dhlab-basel/Knora#926]
- The processing logic of Gravsearch queries in `SearchResponderV2` needs to be refactored and better documented so it becomes easier to maintain and extend.

relates to #910 #1014 #1016 #1010

**Comment by tobiasschweizer** [comment]
### Prequery

The purpose of the prequery is to get an ordered collection of results (sort criteria can be submitted by the user, but the result is always deterministic). Also paging is supported. A prequery is a SPARQL SELECT query. 

In a first step, the Gravsearch query's WHERE clause is transformed and the prequery is generated from this result. The transformation of the Gravsearch query's WHERE clause relies on an implementation of the abstract class `AbstractSparqlTransformer` that goes through the statements one by one and transforms them if necessary.

#### AbstractSparqlTransformer

The `AbstractSparqlTransformer` contains members that need some explanation:
-  `mainResourceVariable: Option[QueryVariable]`: the variable the represents the main resource
- `dependentResourceVariables: mutable.Set[QueryVariable]`: a set of variables representing dependent resources used in the Gravsearch query provided by the user. These are also used in the prequery's WHERE clause and need to be registered to obtain the Iris contained in the prequery's result (SELECT clause).
- `dependentResourceVariablesGroupConcat: Set[QueryVariable]`: contains the Iris of dependent resources as a string using a special separator represented as query variables in the Gravsearch query.
- `valueObjectVariables: mutable.Set[QueryVariable]`: a set of variables representing values (value objects) used in the Gravsearch query provided by the user. These are also used in the prequery's WHERE clause and need to be registered to obtain the Iris contained in the prequery's result (SELECT clause).
- `valueObjectVarsGroupConcat: Set[QueryVariable]`: contains the Iris of value objects as a string using a special separator  that matched the given search criteria.
 
Example:

user provided Gravsearch query:

```sparql
PREFIX incunabula: <http://0.0.0.0:3333/ontology/0803/incunabula/simple/v2#>
PREFIX knora-api: <http://api.knora.org/ontology/knora-api/simple/v2#>

    CONSTRUCT {
        ?page knora-api:isMainResource true .

        ?page knora-api:isPartOf ?book .

        ?page incunabula:seqnum ?seqnum .
    } WHERE {

        ?page a incunabula:page .
        ?page a knora-api:Resource .

        ?page knora-api:isPartOf ?book .
        knora-api:isPartOf knora-api:objectType knora-api:Resource .

        ?book a incunabula:book .

        ?page incunabula:seqnum ?seqnum .
        incunabula:seqnum knora-api:objectType xsd:integer .

        FILTER(?seqnum = 10)

        ?seqnum a xsd:integer .

    }
```
generated prequery:

```sparql
 SELECT DISTINCT ?page (GROUP_CONCAT(DISTINCT(?book); SEPARATOR='') AS ?book__Concat) (GROUP_CONCAT(DISTINCT(?seqnum); SEPARATOR='') AS ?seqnum__Concat) (GROUP_CONCAT(DISTINCT(?book__LinkValue); SEPARATOR='') AS ?book__LinkValue__Concat)
 WHERE {
 ?page <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?page <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 ?page <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/0803/incunabula#page> .
 ?page <http://www.knora.org/ontology/knora-base#isPartOf> ?book .
 ?page <http://www.knora.org/ontology/knora-base#isPartOfValue> ?book__LinkValue .
 GRAPH <http://www.ontotext.com/explicit> {
     ?book__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#LinkValue> .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?book__LinkValue <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?book__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#subject> ?page .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?book__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#object> ?book .
 }
 ?book <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?book <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 ?book <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/0803/incunabula#book> .
 ?page <http://www.knora.org/ontology/0803/incunabula#seqnum> ?seqnum .
 GRAPH <http://www.ontotext.com/explicit> {
     ?seqnum <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?seqnum <http://www.knora.org/ontology/knora-base#valueHasInteger> ?seqnum__valueHasInteger .
 }
 FILTER((?seqnum__valueHasInteger = "10"^^<http://www.w3.org/2001/XMLSchema#integer>))
 }
 GROUP BY ?page
 ORDER BY ASC(?page)
 LIMIT 25
```

state of member vars:

- `mainResourceVariable`: `QueryVariable(page)`
- `dependentResourceVariablesConcat`: `Set(QueryVariable(book__Concat))`
- `valueObjectVariablesConcat`: `Set(QueryVariable(seqnum__Concat), QueryVariable(book__LinkValue__Concat))` 

With this information, the variables in the results of the prequery (its SELECT clause) can be accessed to obtain the Iris of dependent resources and value objects that matched the given search criteria.

**Comment by tobiasschweizer** [comment]
### Main Query

**Comment by benjamingeer** [comment]
I don't understand the function `getSelectVariables`. I think I understand that it returns a sequence of `SelectQueryColumn` which are actually instances of `GroupConcat`, and that `GroupConcat` exists because of its `toSparql` method, which generates the `GROUP_CONCAT` syntax in the prequery. But I don't understand why `getSelectVariables` sets the values of `protected var dependentResourceVariablesGroupConcat` and `protected var valueObjectVarsGroupConcat`. Why do we need `dependentResourceVariablesGroupConcat` and `valueObjectVarsGroupConcat` at all, if `getSelectVariables` already returns all the `SELECT` columns? What purpose do they serve after that?

**Comment by benjamingeer** [comment]
Ah, I think I see now, it's because of these:

https://github.com/dhlab-basel/Knora/blob/2ef9c462946e3f1be84582efed34ec98505838dd/webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala#L2605

https://github.com/dhlab-basel/Knora/blob/2ef9c462946e3f1be84582efed34ec98505838dd/webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala#L2650

So `dependentResourceVariablesGroupConcat` and `valueObjectVarsGroupConcat` serve two purposes:

1. To hold `GroupConcat` objects that will generate the `GROUP_CONCAT` syntax in the prequery.
2. To hold the names of the `__Concat` variables for dependent resources and for values, so we can determine which dependent resources and values were returned by the prequery.

**Comment by benjamingeer** [comment]
For me it's really hard to follow the structure of the code from here until the end of the `createMainQuery` method:

https://github.com/dhlab-basel/Knora/blob/2ef9c462946e3f1be84582efed34ec98505838dd/webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala#L2601

The structure of this code is hard to see because it's a giant `if` statement that's 250 lines long, containing a `for` comprehension that's 140 lines long, which contains a recursive function. It would really help me if all this code was broken up into several smaller functions.

What exactly is `queryResultsSeparatedWithFullQueryPath`? What do you mean by "query path"? To me, a path is a particular sequence of nodes in a tree, like a file path, `/home/foo/bar/baz`. Do you mean something like that?

**Comment by tobiasschweizer** [comment]
> The structure of this code is hard to see because it's a giant if statement that's 250 lines long, containing a for comprehension that's 140 lines long, which contains a recursive function. It would really help me if all this code was broken up into several smaller functions.

I agree. This needs to be refactored like suggested by you. The I can also provide better documentation for the single methods.

> What do you mean by "query path"

I mean "graph pattern". In the docs, we use "graph pattern", but I used "query path" before to refer to the same thing. I will refactor these names and adapt the docs so it becomes clear.

**Comment by benjamingeer** [comment]
OK, so now how about my questions from #910:

1. Should `AbstractSparqlTransformer.handleStandoffLinkFunction` add the target resource to `dependentResourceVariables` if it's a variable? I think the answer is yes.
2. Why doesn't this query work, and what needs to be changed in `SearchResponderV2` to make it work?

```sparql
PREFIX knora-api: <http://api.knora.org/ontology/knora-api/v2#>
PREFIX standoff: <http://api.knora.org/ontology/standoff/v2#>
PREFIX beol: <http://0.0.0.0:3333/ontology/0801/beol/v2#>

CONSTRUCT {
    ?letter knora-api:isMainResource true .
    ?letter beol:hasText ?text .
} WHERE {
    ?letter a beol:letter .
    ?letter beol:hasText ?text .
    ?text knora-api:textValueHasStandoff ?standoffLinkTag .
    ?standoffLinkTag a knora-api:StandoffLinkTag .
  
    FILTER knora-api:standoffLink(?letter, ?standoffLinkTag, <http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA>)
    #<http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> a beol:person .
    
}
```

**Comment by tobiasschweizer** [comment]
I suppose the problem is that the presence of `<http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> a beol:person` or `?person a beol:person`. 

If these are present, then the reifications are present. If not, they are absent. I don't understand yet why.  

In https://github.com/dhlab-basel/Knora/pull/910/commits/6e6d2de19f64badbe79ce99800897359e1decc55 I added some output that shows the expected and present Iris.

**Comment by tobiasschweizer** [comment]
I have just noticed a difference: only if you  include the statement `?person a beol:person` you get: 

> dependent resource Iris concat: Set(QueryVariable(person__Concat))

This would mean that `AbstractSparqlTransformer.handleStandoffLinkFunction` would have to include it, yes. If `?person a beol:person` is present, it would already be there. But since it is a set, this should be ok.

**Comment by tobiasschweizer** [comment]
However, I observe a problem when an Iri is submitted and `<http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> a beol:person .` is not present. Then the reification nodes are missing.

**Comment by benjamingeer** [comment]
> Then the reification nodes are missing.

But they're returned by the prequery:

```sparql
SELECT DISTINCT ?letter (GROUP_CONCAT(DISTINCT(?text); SEPARATOR='') AS ?text__Concat) (GROUP_CONCAT(DISTINCT(?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue); SEPARATOR='') AS ?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue__Concat)
WHERE {
?letter <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
GRAPH <http://www.ontotext.com/explicit> {
    ?letter <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
}
?letter <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/0801/beol#letter> .
?letter <http://www.knora.org/ontology/0801/beol#hasText> ?text .
GRAPH <http://www.ontotext.com/explicit> {
    ?text <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
}
?text <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffLinkTag .
?standoffLinkTag <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#StandoffLinkTag> .
GRAPH <http://www.ontotext.com/explicit> {
    ?letter <http://www.knora.org/ontology/knora-base#hasStandoffLinkTo> <http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> .
}
?standoffLinkTag <http://www.knora.org/ontology/knora-base#standoffTagHasLink> <http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> .
?letter <http://www.knora.org/ontology/knora-base#hasStandoffLinkToValue> ?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue .
GRAPH <http://www.ontotext.com/explicit> {
    ?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#LinkValue> .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#subject> ?letter .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?httprdfhchbiblioup0Q0ZzPSLaULC2tlTs1sA__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#object> <http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> .
}
}
GROUP BY ?letter
ORDER BY ASC(?letter)
LIMIT 25
```

Results:

![screen shot 2018-07-06 at 11 43 02](https://user-images.githubusercontent.com/558389/42372140-c2c5feb0-8111-11e8-8886-0e31e2d6e3d2.png)

**Comment by tobiasschweizer** [comment]
This is why they are expected, but they are not present in the results of the main query after permission checking.

Can you see the output? It should say ` all expected value objects present: false`

**Comment by benjamingeer** [comment]
Yes, that's what I don't understand.

**Comment by tobiasschweizer** [comment]
me neither :-) 

maybe we just have to figure out what happens if the statement is `<http://rdfh.ch/biblio/up0Q0ZzPSLaULC2tlTs1sA> a beol:person`

**Comment by benjamingeer** [comment]
I think the only way I could figure out this code is if you refactored it first as we discussed above.

**Comment by tobiasschweizer** [comment]
> I think the only way I could figure out this code is if you refactored it first as we discussed above.

Ok! I will look at the things I will have to do next week and see how much time I can reserve for it.

**Comment by benjamingeer** [comment]
Fixing #953 could be part of this issue.
